qdb/tracer.py

```diff
@@ -450,6 +503,109 @@ def set_quit(self):
+                        value = progn(
+                            code,
+                            self.eval_fn,
+                            stackframe,
+                        )
+                    except QdbPrognEndsInStatement:
+                        # Statements have no value to print.
+                        pass
+                    else:
+                        if pprint:
+                            value = pformat(value)
+                        if repr_fn:
+                            value = repr_fn(value)
+                        print(value)
+            except Exception as e:
+                outexc = type(e).__name__
+                outmsg = self.exception_serializer(e)
+            else:
+                outmsg = (out.getvalue()[:-1] if out.getvalue() and
+                          out.getvalue()[-1] == '\n' else out.getvalue())
+
+        if outexc is not None or outmsg is not None:
+            self.cmd_manager.send_print(code, outexc, outmsg)
+
+        self.update_watchlist()
+
+    def _stack_jump_to(self, index):
+        """
+        Jumps the stack to a specific index.
+        Raises an IndexError if the desired index does not exist.
+        """
+        # Try to jump here first. This could raise an IndexError which will
+        # prevent the tracer's state from being corrupted.
+        self.curframe = self.stack[index][0]
+
+        self.curindex = index
+        self.curframe_locals = self.curframe.f_locals
+        self.update_watchlist()
+
+    def stack_shift_direction(self, direction):
+        """
+        Shifts the stack up or down depending on dirction.
+        If direction is positive, travel up, if direction is negative, travel
+        down. If direction is 0, do nothing.
+        If you cannot shift in the desired direction, an IndexError will be
+        raised.
+        """
+        if direction == 0:
+            return  # nop
+
+        direction = -1 if direction > 0 else 1
+
+        # The substack is a stack were substack[n] is n + 1 frames away from
+        # curframe where we are traveling in the direction we want to shift.
+        if direction < 0:
+            # We are moving UP the stack:
+            # substack is the stack containing all frames above curframe.
+            substack = self.stack[self.curindex:0:direction]
+        else:
+            # We are moving DOWN the stack:
+            # substack is the stack containing all the frames below curframe.
+            substack = self.stack[self.curindex + 1:]
+
+        if not substack:
+            # If substack is empty, you are at the end of the stack, shifting
+            # in the desired direction is impossible.
+            raise IndexError('Shifted off the stack')
+
+        # Count the number of frames that we are not allowed to stop in.
+        # We add one at the end because there is an implied shift of at least
+        # one stackframe.
+        skip_fn = self.skip_fn
+        diff = sum(1 for _ in takewhile(
```

Comment:
Can we pull this into a function and call it "lazy_len" then (assuming you still prefer this to my suggestion above)?  Having to parse this in line makes it that much harder for me to understand what this function does.